suite/sanity/install.go

```diff
@@ -15,8 +19,22 @@ type installParam struct {
 }
 
 type scriptParam struct {
-	Url  string   `json:"url" validate:"required"`
-	Args []string `json:"args"`
+	Url     string         `json:"url" validate:"required"`
+	Args    []string       `json:"args"`
+	Timeout config.Timeout `json:"timeout"`
+}
+
+// a hack to set default values when deserializing optional/nil-able structs
+// in this case, if the user doesn't specify a script timeout, prefer a non-zero timeout
+func (p *scriptParam) UnmarshalJSON(data []byte) error {
+	type alias scriptParam // to prevent Unmarshal recursion
+	defaults := alias{Timeout: config.Timeout{time.Minute * 5}}
```

Comment:
You don't have to implement `json.Unmarshaler` to be able to control default value although it's tempting.
After tracing the code a bit, I think it would be more idiomatic to have the param struct values implement an interface, 
which you could type switch on in lib/config#Validate, something along the lines of:
```go
func Validate(param interface{}) error {
	err := validator.New().Struct(param)
	if err != nil {
		return trace.Wrap(err)
	}
	if validator, ok := param.(paramValidator); ok {
		return trace.Wrap(validator.CheckAndSetDefaults())
	}
	return nil
}

type paramValidator interface {
	CheckAndSetDefaults() error
}
```
Then in suite/sanity/install.go, for example:
```go
// checkAndSetDefaults verifies this object and sets defaults
func (r *installParam) checkAndSetDefaults() error {
	if r.Script == nil {
		return nil
	}
	return r.Script.checkAndSetDefaults()
}
// ...
// checkAndSetDefaults verifies this object and sets defaults
// implements lib/config#paramValidator
func (r *scriptParam) CheckAndSetDefaults() error {
	if r.Timeout.Duration == 0 {
		r.Timeout.Duration = 5 * time.Minute
	}
	return nil
}
```

Reply:
This seems much cleaner.  Thanks for the suggestion!  I've got a proposal at 585ede6 , though the refactoring became a bit more invasive to accommodate pointers to reflected objects.